FF113 RTCPeerConnectionStats - add docs (#26661)
mdn/content: files/en-us/web/api/rtcpeerconnectionstats/index.md

---
title: RTCPeerConnectionStats
slug: Web/API/RTCPeerConnectionStats
page-type: web-api-interface
browser-compat: api.RTCStatsReport.type_peer-connection
---

{{APIRef("WebRTC")}}

> The [WebRTC API](/en-US/docs/Web/API/WebRTC_API)'s **`RTCPeerConnectionStats`** dictionary provides information about the high level peer connection ({{domxref("RTCPeerConnection")}}).

Reviewer: We usually have "The thing interface(dictionary) of the XYZ API": https://developer.mozilla.org/en-US/docs/Web/API/Response.

```suggestion
The **`RTCPeerConnectionStats`** dictionary of the [WebRTC API](/en-US/docs/Web/API/WebRTC_API) provides information about the high level peer connection ({{domxref("RTCPeerConnection")}}).
```

In particular, it provides the number of unique data channels that have been opened, and the number of opened channels that have been closed.
This allows the current number of open channels to be calculated.

These statistics can be obtained by iterating the {{domxref("RTCStatsReport")}} returned by {{domxref("RTCPeerConnection.getStats()")}} until you find a report with the [`type`](#type) of `peer-connection`.

## Instance properties

- {{domxref("RTCPeerConnectionStats.dataChannelsOpened", "dataChannelsOpened")}}
  - : A positive integer value indicating the number of unique {{domxref("RTCDataChannel")}} objects that have entered the [`open`](/en-US/docs/Web/API/RTCDataChannel/readyState#open) state during their lifetime.
- {{domxref("RTCPeerConnectionStats.dataChannelsClosed", "dataChannelsClosed")}}

>   - : A positive integer value indicating the number of unique {{domxref("RTCDataChannel")}} objects that have left the [`open`](/en-US/docs/Web/API/RTCDataChannel/readyState#open) state during their lifetime (channels that transition to "closing" or "closed" without ever being "open" are not counted in this number).

Reviewer: Should `"open"`, `"closing"` and `"closed"` also be code format? And should `"closing"` and `"closed"` be hyperlinked?

Author: Yes. I will do that.

Author: ```suggestion
  - : A positive integer value indicating the number of unique {{domxref("RTCDataChannel")}} objects that have left the [`open`](/en-US/docs/Web/API/RTCDataChannel/readyState#open) state during their lifetime (channels that transition to [`closing`](/en-US/docs/Web/API/RTCDataChannel/readyState#closing) or [`closed`](/en-US/docs/Web/API/RTCDataChannel/readyState#closed) without ever being `open` are not counted in this number).
```

>     A channel will leave the open state if either end of the connection or the underlying transport is closed.

Reviewer: ```suggestion
    A channel will leave the `open` state if either end of the connection or the underlying transport is closed.
```

### Common instance properties

> The following properties are common to all statistics objects.

Reviewer: ```suggestion
The following properties are common to all WebRTC statistics objects.
```

<!-- RTCStats -->

- {{domxref("RTCPeerConnectionStats.id", "id")}}
  - : A string that uniquely identifies the object that is being monitored to produce this set of statistics.
- {{domxref("RTCPeerConnectionStats.timestamp", "timestamp")}}
  - : A {{domxref("DOMHighResTimeStamp")}} object indicating the time at which the sample was taken for this statistics object.
- {{domxref("RTCPeerConnectionStats.type", "type")}}

>   - : A string with the value `peer-connection`, indicating the type of statistics that the object contains.

Reviewer: ```suggestion
  - : A string with the value `"peer-connection"`, indicating the type of statistics that the object contains.
```

(can't remember what we decided here)

Author: We didn't decide anything, other than this should be consistent within a doc. I'm happy enough either way.

## Examples

This example shows a function to return the total number of open connections, or `null` if no statistics are provided.
This might be called in a loop, similar to the approach used in [`RTCPeerConnection.getStats()` example](/en-US/docs/Web/API/RTCPeerConnection/getStats#examples)

> The method waits on a waits on a {{domxref("RTCPeerConnection")}} for statistics and then iterates the returned {{domxref("RTCStatsReport")}} to get just the stats of type `peer-connection`.

Reviewer: ```suggestion
The function waits for the result of a call to {{domxref("RTCPeerConnection.getStats()")}} and then iterates the returned {{domxref("RTCStatsReport")}} to get just the stats of type `"peer-connection"`.
```

Reviewer: I'm a bit puzzled though. Since we use the `type` property to find the peer connection stats, why is `RTCStatsReport` a map? Why isn't it just an array?

Author: @wbamberg `RTCStatsReport` is an iterator - I don't think we know anything much about the object other than that. You iterate it, checking `type` until it matches what you want. When you have that type, you can query the object then has the properties of that type. I might be missing something.

Reviewer: It's a map:

> The [getStats](https://www.w3.org/TR/webrtc/#widl-RTCPeerConnection-getStats-Promise-RTCStatsReport--MediaStreamTrack-selector)() method delivers a successful result in the form of an [RTCStatsReport](https://www.w3.org/TR/webrtc/#dom-rtcstatsreport) object. An [RTCStatsReport](https://www.w3.org/TR/webrtc/#dom-rtcstatsreport) object is a map between strings that [id](https://www.w3.org/TR/webrtc/#dom-rtcstats-id)entify the inspected objects (id attribute in [RTCStats](https://www.w3.org/TR/webrtc/#dom-rtcstats) instances), and their corresponding [RTCStats](https://www.w3.org/TR/webrtc/#dom-rtcstats)-derived dictionaries.

(https://www.w3.org/TR/webrtc/#dom-rtcstatsreport)

But at least I understand why it is a map now. It's a map from `id` values to `RTCStats` subclasses, where the `id` value identifies the object about which we are getting stats, and which stays constant between `getStats()` calls, enabling us to cross-reference from one stats report to another.

The example code in https://w3c.github.io/webrtc-pc/#getstats-example helped me.

It then returns the total number of open channels using the data in the report.

```js
async function numberOpenConnections (peerConnection) {

>   const stats = await peerConnection.getStats(/* null */);

Reviewer: Why the commented out parameter?

Author: Shouldn't be there.
```suggestion
  const stats = await peerConnection.getStats();
```

> It then returns the total number of open channels using the data in the report.

Author: ```suggestion
It then returns the total number of open channels, or `null`, using the data in the report.
```

  let peerConnectionStats = null;

  stats.forEach((report) => {
    if (report.type === "peer-connection") {
      peerConnectionStats = report;
      break;
    }
  });

>   return peerConnectionStats.dataChannelsOpened - peerConnectionStats.dataChannelsClosed;

Reviewer: This will also throw if there wasn't an instance of this type (and in this case `?.` does not help).

Author: @wbamberg How about?:

```suggestion
result = (peerConnectionStats.dataChannelsOpened && peerConnectionStats.dataChannelsClosed) ?peerConnectionStats.dataChannelsOpened - peerConnectionStats.dataChannelsClosed : null;
return result
```
